### PR TITLE
Apf fix watch panic handling

### LINKED ISSUE #127048: Failure cluster [ae126d38...] `TestApfWatchPanic` flakes a lot
image

from https://storage.googleapis.com/k8s-triage/index.html?pr=1&text=TestApfWatchPanic

##### Error text:
```
Failed;
=== RUN   TestApfWatchPanic
    priority-and-fairness_test.go:187: the APF handler has finished, checking atomicReadOnlyExecuting
    priority-and-fairness_test.go:189: Wanted 0 requests executing, got 1
--- FAIL: TestApfWatchPanic (0.00s)
;
```
#### Recent failures:
[8/31/2024, 6:53:39 PM pr:pull-kubernetes-unit](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/127033/pull-kubernetes-unit/1830015921691824128)
[8/31/2024, 6:37:12 PM ci-kubernetes-unit](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-unit/1830011786368126976)
[8/31/2024, 5:37:12 PM ci-kubernetes-unit](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-unit/1829996686651428864)
[8/31/2024, 4:43:26 PM pr:pull-kubernetes-unit](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/127033/pull-kubernetes-unit/1829983138311311360)
[8/31/2024, 4:38:54 PM pr:pull-kubernetes-unit](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/127046/pull-kubernetes-unit/1829982003957600256)

/kind failing-test

### PR DESCRIPTION
**Pull request by MikeSpreitzer**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
This PR (a) introduces and uses controlflow.TryFinally to make the control flow clearer and (b) simplifying and clarifying the signalling between the goroutine that the APF handler forks in the WATCH case and the main goroutine while also adding handling of the case of both goroutines panicking. Part (b) is structured as two commits: the first removes the relaying of a panic, the second restores the relay and handles all four panic/no-panic cases.

#### Which issue(s) this PR fixes:

Fixes #127048 

#### Special notes for your reviewer:
For a version without introducing and using `controlflow.TryFinally`, see #127762.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by MikeSpreitzer** [comment]
The deltas shown by GitHub may be a little hard to follow, but if you look at the revised code I think that you will see that using controlflow.TryFinally makes it easier to follow (making the order of the source agree with the order of execution).

File: staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go [hunk 1]
```diff
@@ -31,6 +31,7 @@ import (
 	epmetrics "k8s.io/apiserver/pkg/endpoints/metrics"
 	apirequest "k8s.io/apiserver/pkg/endpoints/request"
 	"k8s.io/apiserver/pkg/server/httplog"
+	cflow "k8s.io/apiserver/pkg/util/controlflow"
```
Comment (neolit123) [inline staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go:34]: this could be just 
```suggestion
	"k8s.io/apiserver/pkg/util/controlflow"
```

File: staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go [hunk 2]
```diff
@@ -264,24 +251,48 @@ func (h *priorityAndFairnessHandler) Handle(w http.ResponseWriter, r *http.Reque
 			h.fcIfc.Handle(handleCtx, digest, noteFn, estimateWork, queueNote, execute)
 		}()
 
-		select {
-		case <-shouldStartWatchCh:
-			watchCtx := utilflowcontrol.WithInitializationSignal(ctx, watchInitializationSignal)
-			watchReq = r.WithContext(watchCtx)
-			h.handler.ServeHTTP(w, watchReq)
+		cflow.TryFinally(func() {
```
Comment (neolit123) [inline staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go:254]: using TryFinally instead of the defer SGTM, but i don't understand the APF logic, so leaving that to the maintainers to review.

**Comment by seans3** [comment]
Please have a look at PR: https://github.com/kubernetes/kubernetes/pull/127089 which is also associated with the issue.

File: staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go (hunk 1 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go]: You're right. I have made that change.

**Comment by MikeSpreitzer** [comment]
My recent updates are explained at https://github.com/kubernetes/kubernetes/pull/127029#issuecomment-2348219023 .

**Comment by MikeSpreitzer** [comment]
@MikeSpreitzer

**Comment by tkashem** [comment]
Can we address these concerns in separate PRs:
- a) the refactor only try-finally pattern, it's across multiple packages, this can go in a separate PR. I personally think it does not make the go code idiomatic, inline func with defer seems to be idiomatic and a prevalent style. However, if folks like this pattern I am fine with it (an alternate is to refactor out the standard and watch handler via a strategy, this might make the code easier to read).
- b) the bookkeeping issue warrants a separate PR in my opinion, the issue is reproduced, and the fix is backed by `stress` test result. If by any chance it introduces other flake, it can be reverted atomically. 
- c) the other changes in this PR, looks to be refactor only, if it is in a separate PR it would be easier to review 

I think these can be done in any order, thoughts?

**Comment by MikeSpreitzer** [comment]
I just realized that I had forgotten to update the opening comment of this PR when I added the latest commits. I have done that now.

@tkashem: I think there are two major changes in this PR, as now outlined in the opening comment, and they could indeed be done in separate PRs.

**Comment by MikeSpreitzer** [comment]
The force-push to aaa2cf8f11e simplifies the func that closes `decidedCh`: because only one goroutine invokes it, a simple boolean can be used instead of a `sync.Once`.

**Comment by MikeSpreitzer** [comment]
The most recent push adds a commit that restores relaying of a panic from the wait goroutine. The code in that goroutine is not trivial, so a panic is a possibility to cover. The relaying is so that it is handled in the apiserver's normal way (by `WithPanicRecovery`). Hopefully the revised code structure is clearer.

File: staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go [hunk 3]
```diff
@@ -264,24 +249,40 @@ func (h *priorityAndFairnessHandler) Handle(w http.ResponseWriter, r *http.Reque
...
+					combined := fmt.Sprintf("wait and watch both panicked.\nwait: %s\n\nwatch: %v", waitReport, recoveredFromWatch)
+					panic(combined)
+				} else {
+					// Experiment shows that this preserves the stack, so faithfully that both panics show up.
+					panic(recoveredFromWatch)
+				}
+			} else if waitPanicked {
+				panic("APF wait goroutine panicked: " + waitReport)
+			}
+		})
```
[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go:285]: Yes, but I think it is more complicated than necessary. Note that `panic.go` adds more lines than all of this PR (which also removes more lines than #127029).

If we wanted, we could collect 273--284 here into a func and offer it for general use. That would be simple and concentrate that concern outside this func.

**Comment by MikeSpreitzer** [comment]
The force-push to 3cfc7a708ad squashes out an uninteresting commit.

**Comment by MikeSpreitzer** [comment]
The force-push to 120a5e964c2 is a rebase onto `master`.
